libgalois/src/analytics/cdlp/cdlp.cpp

```diff
@@ -0,0 +1,402 @@
+    return pg_result.error();
+  }
+  auto graph = pg_result.value();
+
+  using BiDirGraphView = katana::TypedPropertyGraphView<
+      katana::PropertyGraphViews::BiDirectional, typename Algorithm::NodeData,
+      typename Algorithm::EdgeData>;
+
+  auto bidir_view =
+      KATANA_CHECKED(BiDirGraphView::Make(pg, {output_property_name}, {}));
+
+  Algorithm algo(plan);
+
+  algo.Initialize(&graph);
+
+  katana::StatTimer execTime("CDLP");
+
+  execTime.start();
+  algo(&graph, bidir_view, max_iterations);
+  execTime.stop();
+
+  algo.Deallocate(&graph);
+  return katana::ResultSuccess();
+}
+
+katana::Result<void>
+katana::analytics::Cdlp(
+    PropertyGraph* pg, const std::string& output_property_name,
+    size_t max_iterations, CdlpPlan plan) {
+  switch (plan.algorithm()) {
+  case CdlpPlan::kSynchronous:
+    return CdlpWithWrap<CdlpSynchronousAlgo>(
+        pg, output_property_name, plan, max_iterations);
+  /// TODO: Asynchronous Algorithm will be implemented later after Synchronous
+  /// is done for both shared and distributed versions.
+  /*
+  case CdlpPlan::kAsynchronous:
+    return CdlpWithWrap<CdlpAsynchronousAlgo>(
+        pg, output_property_name, plan, max_iterations);
+  */
+  default:
+    return ErrorCode::InvalidArgument;
+  }
+}
+
+katana::Result<void>
+katana::analytics::CdlpAssertValid(
+    PropertyGraph* pg, const std::string& property_name) {
+  using CommunityType = uint64_t;
+  struct NodeCommunity : public katana::PODProperty<CommunityType> {};
+
+  using NodeData = std::tuple<NodeCommunity>;
+  using EdgeData = std::tuple<>;
+  typedef katana::TypedPropertyGraph<NodeData, EdgeData> Graph;
+  typedef typename Graph::Node GNode;
+
+  auto pg_result = Graph::Make(pg, {property_name}, {});
+  if (!pg_result) {
+    return pg_result.error();
+  }
+
+  auto graph = pg_result.value();
+
+  using BiDirGraphView = katana::TypedPropertyGraphView<
+      katana::PropertyGraphViews::BiDirectional, NodeData, EdgeData>;
+
+  auto bidir_view =
+      KATANA_CHECKED(BiDirGraphView::Make(pg, {property_name}, {}));
+
+  auto is_bad = [&graph, &bidir_view](const GNode& node) {
+    const auto ndata = graph.template GetData<NodeCommunity>(node);
+    typedef boost::unordered_map<CommunityType, size_t> Histogram_type;
+    Histogram_type histogram;
+    // Incoming edges
+    for (auto e : bidir_view.in_edges(node)) {
+      auto src = bidir_view.in_edge_dest(e);
+      const auto sdata = graph.template GetData<NodeCommunity>(src);
+      histogram[sdata]++;
+    }
+
+    // Outgoing edges
+    for (auto e : bidir_view.edges(node)) {
+      auto dest = bidir_view.edge_dest(e);
+      const auto ddata = graph.template GetData<NodeCommunity>(dest);
+      histogram[ddata]++;
+    }
+
+    /// Pick the most frequent communtiy for node
+    /// Pick the smallest one if more than one max frequent exist.
+    /// TODO: This needs to be fix for the oscillation of the label cases
+    /// It returns Failed for those cases.
```

Comment:
"From our experiments, we found that irrespective of n (# of nodes in the graph), 95% of the nodes or more are classified correctly by the end of iteration 5."

I think this percentage should be considered in CdlpAssertValid; in a way that if it is more than a specific threshold it returns true. Right now, it fails when oscillations happen; since in that case, there are some nodes that have a label that is not the most frequent one in their immediate neighborhood.

I think we should add this percentage to our statistics as well; so based on this percentage of accurate node classification one can alter the number of iterations; however, from some point on, increasing the number of iteration does not change this percentage due to the label oscillation. Concerns: Computing this percentage can be expensive. Maybe, we can do it on user demand.